evap/static/ts/src/csrf-utils.ts

```diff
@@ -13,22 +13,50 @@ function getCookie(name: string): string | null {
 }
 const csrftoken = getCookie("csrftoken")!;
 
-function isMethodCsrfSafe(method: string): boolean {
+function doesMethodRequireCsrfToken(method: string): boolean {
     // these HTTP methods do not require CSRF protection
-    return ["GET", "HEAD", "OPTIONS", "TRACE"].includes(method);
+    return ["POST", "PUT", "DELETE", "PATCH"].includes(method.toUpperCase());
+}
+
+function isUrlCrossDomain(url: string): boolean {
+    return new URL(document.baseURI).origin !== new URL(url, document.baseURI).origin;
 }
 
 // setup ajax sending csrf token
 $.ajaxSetup({
     beforeSend: function (xhr: JQuery.jqXHR, settings: JQuery.AjaxSettings) {
-        const isMethodSafe = settings.method && isMethodCsrfSafe(settings.method);
-        if (!isMethodSafe && !this.crossDomain) {
+        const methodRequiresCsrfToken = settings.type && doesMethodRequireCsrfToken(settings.type);
+        if (methodRequiresCsrfToken && !this.crossDomain) {
             xhr.setRequestHeader("X-CSRFToken", csrftoken);
         }
     },
 });
 
+// Add CSRF token to regular forms on submit.
+// This reduces boilerplate code in the templates and makes the form's HTML cacheable.
+for (const form of document.forms) {
+    form.addEventListener("submit", (ev: Event) => {
+        // in case a form gets submitted twice
+        if (form.querySelectorAll("input[name=csrfmiddlewaretoken]").length > 0) {
+            return;
+        }
+
+        const { submitter } = ev as SubmitEvent;
+        const action = submitter?.getAttribute("formaction") ?? form.getAttribute("action");
+        const method = submitter?.getAttribute("formmethod") ?? form.getAttribute("method");
```

Comment:
Consider the following scenario:
1. There is a form with multiple submit-inputs, one of which has `formmethod=post`, and another with `formmethod=get`
2. First, we submit with the post method which inserts the token
3. the server responds with http 204 and the site doesn't reload (or it doesn't reload for any other reason)
4. now, the token is in the form and subsequent submits with the get-input will contain the token :(

The same problem occurs with `formaction` which eliminates the `isUrlCrossDomain` check on subsequent calls.

@He3lixxx and I weren't able to come up with a satisfying solution that fits into the current model. Richard identified that this currently does two things when we really only cared about one:

1. Insert the _value_ of the token dynamically
2. Insert the _field_ for the token dynamically

If we went back to having a `{{ evap_dynamic_csrf }}` or so, we could still insert the value dynamically, but would still have the responsibility for deciding whether or not to include csrf tokens in the form code itself. This is analogous to the responsibilities when using the django csrf tag.

Reply:
Doesn't the same problem occur with the previously used {{csrf_token}} as well?

I don't think we need to support this. I can try and add an assert to catch this if you want.

I don't think the cacheability alone is worth the added complexity of this PR. So the point of this PR *is* to insert everything dynamically for easier development and to avoid the boilerplate code.

Reply:
Hmm, tough situation.

> Doesn't the same problem occur with the previously used {{csrf_token}} as well?

My thoughts were:
Any form that mixes POST and GET is weird. Any form that has external or mixed endpoints/actions is weird. I don't see how we'd ever need something like that. However, if someone opened a PR with a good reason, we might accept it without thinking about security implications. Now, if the PR contains the `{{csrf_token}}` in this form, we'd be more likely to think about it compared to if the token was added silently.
(One could argue that a piece of javascript that changes a form's action or method has the same issue, though.)

> I don't think the cacheability alone is worth the added complexity of this PR. So the point of this PR is to insert everything dynamically for easier development and to avoid the boilerplate code.

Easier Development is usually better, but in this case, I'd say it gets us closer to doing security stuff wrong, so personally, I wouldn't do it -- I feel like this PR shows how much we're playing with fire here.

(For #1645, a clear_navbar_cache call at the right location should also work to fix the issue for now)

Reply:
I tried the formdata event. The cool thing is that the FormData is constructed from scratch for each request, so submitting a form multiple times is not an issue anymore, and no html element construction is necessary anymore. I still had to use the submit event to get the submitter, though.

To put the "playing with fire" into perspective: The actual CSRF verification is done on the server and unchanged, so this PR cannot make us vulnerable to missing tokens. Regarding leaking the token somewhere where it shouldn't be leaked: Consider that it is already stored in the cookie, which is not an uber secure place to begin with. I doubt it would make anything considerably more insecure if we e.g. accidentally put the CSRF token into a GET request (though I do believe I have made that pretty much impossible with my recent change).

@He3lixxx and/or @niklasmohrin can you please make a call? If you give me a thumbs up, I'd summarize all the edge cases we found into a comment before merging. Otherwise I'd close this, which is also fine.